src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1715,9 +1716,27 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
             var op1 = BindValue(node.Left, diagnostics, BindValueKind.Assignment); // , BIND_MEMBERSET);
             var op2 = BindValue(node.Right, diagnostics, BindValueKind.RValue); // , BIND_RVALUEREQUIRED);
 
+            if (op1.Kind == BoundKind.DiscardedExpression)
+            {
+                op1 = InferTypeForDiscard((BoundDiscardedExpression)op1, op2, diagnostics);
+            }
+
             return BindAssignment(node, op1, op2, diagnostics);
         }
 
+        private static BoundExpression InferTypeForDiscard(BoundDiscardedExpression op1, BoundExpression op2, DiagnosticBag diagnostics)
+        {
+            if (op2.Type == null)
+            {
+                Error(diagnostics, ErrorCode.ERR_DiscardVariableAssignedBadValue, op1.Syntax, op2.Syntax);
```

Comment:
>op2.Syntax [](start = 94, length = 10)

We usually don't use syntax as a diagnostics argument, it might have line breaks, comments, etc. Consider using op2.Display instead.